### ISSUE TITLE
Create a printable Status Report

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#5821]
## Description of the need

In https://github.com/backdrop/backdrop-issues/issues/5791 we are looking into providing a list of enabled modules. This however was in response to a perceived need for site information for debugging. 

As in Q: "My site is producing fatal errors" A: "What modules you have installed?" 

But I find that frequently other questions accompany, like "What Backdrop/PHP/MySQL version, which theme are you using, is Javascript aggregated, etc etc"

Noting that the Status Report page provides a lot of that info already, but is not printable, and may be too much info in some situations

## Proposed solution

A link/button on the status Report page to print a report. I propose a radio with two options: "full", and "Basic" (or maybe "Safe"). When you click the button you get the Status report info in printable format, with added stuff like enabled/disabled modules, themes, layout, etc etc. "Full" prints everything on the Status Report page, "Basic/Safe" would be default, and print only stuff that shouldn't put your site at risk if you post it in the Forums.

An example of a similar effort for WP is this plugin: https://wordpress.org/plugins/website-information-report/

## Alternatives that have been considered

A View of enabled modules alone is being considered for core in https://github.com/backdrop/backdrop-issues/issues/5791

**Comment by indigoxela** [comment]
@docwilmot some questions: where to put it and what should it look like. And does "print" mean "export to copy/paste", or does it mean "download a file"?

I see several use-cases for such an export.

- In the forum, so people can more easily compare installs without asking back 5 times
- In our issue tracker, possibly even sort of mandatory for opening a bug report
- As a possible admin tool to export a "site status" to compare later, what changed since last report (and what's a candidate for the current problem)

However, one thing seems important to me: under *no* circumstances this export should ever contain sensitive data not appropriate for the public. Personally a user count (or even worse, status) would be one of those sensitive data sets.

**Comment by docwilmot** [comment]
> @docwilmot some questions: where to put it and what should it look like. 

Maybe an operations fieldset? Like this?
![printable](https://user-images.githubusercontent.com/1042283/194722716-59931e8d-8791-4e2c-acbb-0c2fe59afb74.png)

Or maybe just a link at the top?

> And does "print" mean "export to copy/paste", or does it mean "download a file"?

I'm thinking just generates a page with just plain text without layout, generated on a new window, that user can CTRL-A/CTRL-C
Like:

**Backdrop CMS**
1.24.x-dev
**MySQL Database**
MySQL, MariaDB, or equivalent version 5.7.31
**PHP**
Version: 7.4.9 (PHP information)
**Enabled modules**
Field
Layout
Backup and migrate
etc

**Comment by indigoxela** [comment]
@docwilmot hm, some head scratching here, when looking at your mockup: "Basic report hides sensitive data" - that's the opposite of my suggestion to never ever include sensitive data.

Why do you want to include sensitive data, anyway?

**Comment by docwilmot** [comment]
I dont particularly *want* anything, just making suggestions, as you say its a mockup. You may not like one idea, but others might. We'll decide by the end of this I suspect.

But the practical reason though why we should *consider* this, if you put a print button on a page, it may seem strange that it doesnt actually allow you to print the page but instead you can only print bits of the page. If we decide a full print is unacceptable, then perhaps we should consider a separate menu item for a printable report, instaed of any link/button on the Status Report page. I use the word "print" here of course to mean "generate a printable report" or such.

Again, open to discussion.

**Comment by olafgrabienski** [comment]
> An example of a similar effort for WP is this plugin: https://wordpress.org/plugins/website-information-report

For reference, the information collected by the WordPress plugin:

- PHP version
- WordPress version

Numbers:

- User accounts
- Published posts
- Published / approved / pending comments
- Active / inactive plugins

**Comment by olafgrabienski** [comment]
> For reference, the information collected by the WordPress plugin ...

While the numbers of users, posts, and comments are usually not very interesting for Backdrop troubleshooting, PHP version, CMS version and plugins (modules) are. Which information else would be interesting for our use case (provide helpful information for troubleshooting)?

**Comment by jenlampton** [comment]
>  propose a radio with two options: "full", and "Basic" (or maybe "Safe"). When you click the button you get the Status report info in printable format, 

I dislike the idea of adding a button that does the same thing as printing. Everybody already has a way to print :) I would prefer if we either added a print style sheet to the current status report page, or even generate something else like a json feed of all the same data.

> with added stuff like enabled/disabled modules, themes, layout, etc etc. "Full" prints everything on the Status Report page, "Basic/Safe" would be default, and print only stuff that shouldn't put your site at risk if you post it in the Forums.

I also dislike the idea of hiding important stuff from display. If the list of modules is important, it should be somewhere that people can see it too (rather than only being available for print, or in a json export)

Can this issue become "Add a print style sheet to the status report", and we can follow up with other things people would like to see on the status report?

**Comment by klonos** [comment]
I think that the mentioning of the word "print" is confusing here. Unless I misunderstood, I think what @docwilmot meant was a "plain text" thingy that could easily be copy/pasted and/or possibly downloaded as a .txt file (that can then be uploaded to the forum or the issue queue). So lets not linger on that word/functionality please.

With the above in mind, I think that if we had something similar to the single config export text box, it would be enough. Then perhaps also something like a copy-to-clipboard link/button (which IIRC @cellear has requested to be added in the config export text boxes some time ago).

**Comment by docwilmot** [comment]
> Can this issue become "Add a print style sheet to the status report", and we can follow up with other things people would like to see on the status report?

I dont think that accomplishes the issue though. What we're asking for here is a *prinatble* report that one can copy paste easily. A print style sheet would need to be printed to PDF then screenshotted, then pasted.

This is of course something that screams "contrib!" but to have to download a module just so you can print your module list and PHP version seems excessive.

How about a menu link that says "Debug report" that does what we need?

**Comment by bugfolder** [comment]
> I think that the mentioning of the word "print" is confusing here. Unless I misunderstood, I think what @docwilmot meant was a "plain text" thingy that could easily be copy/pasted...

So how about "Generate plain text report"?

**Comment by kiamlaluno** [comment]
When I read *printable*, I usually think of something that gives less problems when printed, not to something that can be printed. In most of the cases, as developer, I take that means the plain text version of a page that normally is shown with images and styled boxes.

The WordPress plugin linked in the issue summary doesn't provide a plain text report, though.

![screenshot](https://user-images.githubusercontent.com/29851/210631980-6890d173-d717-4f60-ba1a-a524e4cf8e48.png)

The part listing the installed plugins is probably the part that is more problematic, if I want to copy it as text.

![screenshot](https://user-images.githubusercontent.com/29851/210632752-90c79cda-d734-41ac-98a0-79de464ac12a.png)

I also wonder what exactly this feature request is requesting and what use case is trying to cover.

If the need is a page that reports information that can be easily copy-pasted in a post, where I could report a bug for which I am asked for more information about my site, it's sufficient a page where the information is shown without being formatted in multiple columns, as in the following text. (The text is a simplified version of what information is really shown.)

**Backdrop version**: 1.23.1
**Installed modules**: Administration Bar (1.23.1), Color (1.23.1), Entity UI (1.x-1.0.1).

**Comment by docwilmot** [comment]
Perhaps we should just add this to the Devel module "Information" menu instead. Most people who need help debugging might benefit from having Devel installed at least temporarily while being helped anyway. And that would fulfill the "test in contrib first" principle.

**Comment by kiamlaluno** [comment]
I think this can be a Backdrop core feature, but I also understand it's more suitable for the Devel module.
I guess it depends on the amount of information I could need to provide to help understand what the issue is. The Devel module is the perfect module to get that type of information, but it could also provide too much information in some cases.

If the decision is to implement this feature in the Devel module, I am fine with that too.

**Comment by stpaultim** [comment]
I THINK that the problem we're trying to solve here, is to make it easy to cut and paste important information that might help a person debug problems with their BackdropCMS website. Here is what happens if I cut and paste the status report from a demo site I just created. 

Wow, Github handled this better than I expected. 

- - - - - - - - - 

Google Analytics module | Not configured lessGoogle Analytics module has not been configured yet. Please configure its settings from the Google Analytics settings page.
-- | --
Warning | Trusted Host Setting | Not configured lessThe trusted_host_patterns setting is not configured in settings.php. It is highly recommended that you configure this to protect your site. See Protecting against HTTP HOST Header attacks for more information.
Info | Backdrop CMS | 1.24.x-dev
Info | Backdrop CMS core update status | No update data available more
Info | MySQL Database | MySQL, MariaDB, or equivalent version 5.5.5-10.3.27-MariaDB
Info | MySQL Database 4-byte UTF-8 support | Not enabled more
Info | PHP | Version: 7.4.33 (PHP information)
Info | Telemetry | Enabled more
Info | Web server | Apache/2.4.54 (Debian)
OK | Access to update.php | Protected
OK | Backdrop CMS update notifications | Enabled
OK | Cron maintenance tasks | Last run 1 hour 1 min ago (Run cron manually) more
OK | Database updates | Up to date
OK | Display error messages | Disabled more
OK | File system | Writable (public download method)
OK | GD library rotate and desaturate effects | bundled (2.1.0 compatible)
OK | jQuery | jQuery version 1.12.4 / jQuery UI version 1.13.0
OK | Node Access Permissions | Disabled (Rebuild permissions) more
OK | PHP max_input_vars setting | PHP limits the number of input variables in each form to 10,007 items.
OK | PHP extensions | Enabled: date, dom, filter, gd, hash, json, pcre, pdo, session, SimpleXML, SPL, xml
OK | PHP memory limit | 1G
OK | PHP register globals | Disabled
OK | Unicode library | PHP Mbstring Extension
OK | Upload progress | Client-side upload progress enabled

Google Analytics module
Not configured [less](http://enabled-modules.lndo.site/admin/reports/status#)
Google Analytics module has not been configured yet. Please configure its settings from the [Google Analytics settings page](http://enabled-modules.lndo.site/admin/config/system/googleanalytics).
Warning
Trusted Host Setting
Not configured [less](http://enabled-modules.lndo.site/admin/reports/status#)
The trusted_host_patterns setting is not configured in settings.php. It is highly recommended that you configure this to protect your site. See [Protecting against HTTP HOST Header attacks](https://docs.backdropcms.org/documentation/trusted-host-settings) for more information.
Info
Backdrop CMS
1.24.x-dev
Info
Backdrop CMS core update status
No update data available [more](http://enabled-modules.lndo.site/admin/reports/status#)
Info
MySQL Database
MySQL, MariaDB, or equivalent version 5.5.5-10.3.27-MariaDB
Info
MySQL Database 4-byte UTF-8 support
Not enabled [more](http://enabled-modules.lndo.site/admin/reports/status#)
Info
PHP
Version: 7.4.33 ([PHP information](http://enabled-modules.lndo.site/admin/reports/status/php))
Info
Telemetry
Enabled [more](http://enabled-modules.lndo.site/admin/reports/status#)
Info
Web server
Apache/2.4.54 (Debian)
OK
Access to update.php
Protected
OK
Backdrop CMS update notifications
Enabled
OK
Cron maintenance tasks
Last run 1 hour 1 min ago ([Run cron manually](http://enabled-modules.lndo.site/admin/reports/status/run-cron)) [more](http://enabled-modules.lndo.site/admin/reports/status#)
OK
Database updates
Up to date
OK
Display error messages
Disabled [more](http://enabled-modules.lndo.site/admin/reports/status#)
OK
File system
Writable (public download method)
OK
GD library rotate and desaturate effects
bundled (2.1.0 compatible)
OK
jQuery
jQuery version 1.12.4 / jQuery UI version 1.13.0
OK
Node Access Permissions
Disabled ([Rebuild permissions](http://enabled-modules.lndo.site/admin/reports/status/rebuild)) [more](http://enabled-modules.lndo.site/admin/reports/status#)
OK
PHP max_input_vars setting
PHP limits the number of input variables in each form to 10,007 items.
OK
PHP extensions
Enabled: date, dom, filter, gd, hash, json, pcre, pdo, session, SimpleXML, SPL, xml
OK
PHP memory limit
1G
OK
PHP register globals
Disabled
OK
Unicode library
PHP Mbstring Extension
OK
Upload progress
Client-side upload progress enabled

**Comment by stpaultim** [comment]
This is what it looks like when I cut and paste into the BackdropCMS forum

![image](https://user-images.githubusercontent.com/3144571/210709257-fac8f5a3-b076-4446-8dcc-1b68f4830c19.png)

**Comment by olafgrabienski** [comment]
For reference, below is a copy & paste example of the *Available updates* page (admin/reports/updates). While it's not easy to read in plain text, and core modules are missing, as an admin page it's the perfect fit to get an overview of a site's contrib projects. Note that it mentions release versions as well.

```
Backdrop CMS
Update available
	
Backdrop 1.23.0
Update available
Recommended version: 1.23.1 (2022-Dec-03)

    Download
    Release notes

Latest version: 1.24.0-preview (2023-Jan-04)

    Download
    Release notes

Includes: Administration Bar, Bartik, Basis, Block, Boxton, CKEditor, Configuration Manager, Contact, Contextual Links, Dashboard, Database Logging, Date, Entity, Field, Field SQL Storage, Field UI, File, Filter, Geary, Harris, Image, Language, Layout, Link, List, Locale, Menu, Moscone, Moscone Flipped, Node, Number, Options, Path, Project Installer, Redirect, Rolph, Search, Seven, Simmons, Sutro, System, Taxonomy, Taylor, Taylor Flipped, Telemetry, Text, Update Manager, User, Views, Views UI
Modules
Current
	
AudioField 1.x-1.2.0
Up to date
Current
	
Backup and Migrate 1.x-1.0.21
Up to date
Current
	
CSS Injector 1.x-1.11.1
Up to date
Current
	
Entity Plus 1.x-1.0.17
Up to date
Current
	
Field Group 1.x-1.1.1
Up to date
Current
	
GDPR Cookies 1.x-1.3.2
Up to date
Current
	
Override node options 1.x-1.13.2
Up to date
Current
	
Title Display 1.x-1.2.2
Up to date
Current
	
Video Embed Field 1.x-2.0-alpha
Up to date
Current
	
YouTube Field 1.x-1.0.2
Up to date
Themes
Current
	
Thesis 1.x-1.0.0
Up to date
Depends on: Basis
```

**Comment by klonos** [comment]
in #5791 I have provided some screenshots of a local PoC implementation of a "debug info" page that allows generating various sets of debug info (for starters it's enabled modules, and php/webserver/database info, but can be extended further). In the mockup there are 2 links - one for copying the info in the clipboard, and another for downloading it as a .txt file. ...should I provide that as a PR here instead of #5791?

Although I haven't implemented the copy-to-clipboard functionality yet, I was looking at https://github.com/zenorocha/clipboard.js which is a tiny library (instead of crafting something custom which we'd then need to maintain). I thought that we could add the functionality of copying things on the clipboard to various other places, like error messages and log entries etc.

While working on my PoC I also thought of this: we could add hidden text areas in various pages where we have nice HTML tables of information that could possibly be requested in the issue queue or the forum. Expose the "copy to clipboard" link on the page, right above the visible HTML table. Clicking the link/button would then (transparently to the user) copy a plain-text version of the pretty HTML table (the actual source would be the hidden text area).

**Comment by klonos** [comment]
@docwilmot and other here, please check out my latest PR for https://github.com/backdrop/backdrop-issues/issues/5791. It adds a "Debug information" page under reports, which I have now simplified down to this:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/54df941b-2cde-47e4-9ccf-45f7af0f3b14)

**Comment by indigoxela** [comment]
That looks great! One suggestion: "used when editing or ..." - I think, that's unnecessary and actual misleading info, as the setting to use the admin theme for node create and edit is optional and could be off.

**Comment by olafgrabienski** [comment]
@klonos I like the report as well! Two ideas:

- Add a heading at top of the modules, e.g. "Enabled modules"
- Group modules by origin (core / contrib / custom), or does this make the report already too complex?

---

> "used when editing or ..." - I think, that's unnecessary and actual misleading info, as the setting to use the admin theme for node create and edit is optional and could be off.

Yes, it could be off for the admin theme, but then the report shows Basis as 'content editing' theme:

```
Default theme:             Basis* (basis) [core]
Admin theme:               Seven (seven) [core]
                           *used when editing or creating content
```

Maybe the asterisk isn't clear enough?

**Comment by klonos** [comment]
> as the setting to use the admin theme for node create and edit is optional and could be off.

@indigoxela that checkbox allows to select if the admin theme will be used for content editing. and the default OOTB is that:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/639145ab-f968-44bb-8df2-bc7bb9214d4b)

If you untick the checkbox, what happens is that the default (frontend) theme is being used when editing. My implementation takes that into account, and it indicates which of the two themes is being used for content editing (the admin or the default FE theme). Here's what happens with the default (checkbox ticked):
```
Default theme:             Basis (basis) [core]
Admin theme:               Seven* (seven) [core]
                           *used when editing or creating content
```
...and here's the debug output with that checkbox unticked:
```
Default theme:             Basis* (basis) [core]
Admin theme:               Seven (seven) [core]
                           *used when editing or creating content
```

If the asterisk is hard to notice, I can make it so that this line is directly after the theme that has been configured to be used for content editing. Like so:
```
Default theme:             Basis* (basis) [core]
                           *used when editing or creating content
Admin theme:               Seven (seven) [core]
```

In general though, where that indication helps to save time in back and forth questions during troubleshooting is this:
> - hello, I'm having this issue with my site when trying to edit/save content
> - which themes are you using?
> - theme xyz is my admin theme, and abc is my frontend theme
> - and which of these two is configured to be used when editing content?

Instead of all these questions, we have all the information we need upfront (which is the whole point of this feature 😉 ).

> Add a heading at top of the modules, e.g. "Enabled modules"

Will do that @olafgrabienski 👍🏼 ...stay tuned 😉 

> Group modules by origin (core / contrib / custom), or does this make the report already too complex?

I have tried that in early implementations, but it made it harder to search for modules. Listing modules alphabetically is better I think. I might file a separate PR with your suggestion, just so that we can compare and decide which one works better.

> Nice! Is there a reason for the columns to be so far apart? How will those translate when copying/pasting?

@argiepiano I have consolidated things a bit further, and the columns aren't that far apart now. Please feel free to play with the PR sandbox and experiment with copy/pasting the debug output in a comment in the forum or here on GitHub (or see my reply to @indigoxela above, where I have pasted 3 excerpts of that output). We could certainly improve further, and I'm open to suggestions.

Latest screenshot from the PR sandbox:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/7ea422b4-644f-45e1-aed8-788ed45939d0)

**Comment by klonos** [comment]
@docwilmot should we close this issue here, and focus on #5791 instead, or would you like us to do the same with everything on the site status report? (provide a plain-text debug-report-like version of it)

**Comment by indigoxela** [comment]
Re the theme used for editing content...

> My implementation takes that into account, and it indicates which of the two themes is being used for content editing

Ah, I didn't realize that. Really subtle. :wink: Forget my comment then.

**Comment by docwilmot** [comment]
> @docwilmot should we close this issue here, and focus on #5791 instead, or would you like us to do the same with everything on the site status report? (provide a plain-text debug-report-like version of it)

Looks like these two issues have become almost the same, so I suspect you can close whichever you think is least useful at this stage.

**Comment by klonos** [comment]
Cool. Since all the PRs are on the other issue, I will go ahead and close this one here.

If in the future we decide that we want to explicitly provide a way to copy/paste a printer-friendly version of the site status page (or any other report), then we can reopen/revisit.
